src/charm.py

```diff
@@ -8,18 +8,157 @@
 
-from ops.charm import CharmBase
+from charms.observability_libs.v0.kubernetes_service_patch import KubernetesServicePatch
+from charms.traefik_k8s.v1.ingress import (
+    IngressPerAppReadyEvent,
+    IngressPerAppRequirer,
+    IngressPerAppRevokedEvent,
+)
+from jinja2 import Template
+from ops.charm import CharmBase, PebbleReadyEvent
 from ops.main import main
+from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus, ModelError, WaitingStatus
+from ops.pebble import ChangeError, Layer
 
 logger = logging.getLogger(__name__)
 
+OATHKEEPER_API_PORT = 4456
+
 
 class OathkeeperCharm(CharmBase):
     """Charmed Ory Oathkeeper."""
 
     def __init__(self, *args):
         super().__init__(*args)
 
+        self._container_name = "oathkeeper"
+        self._container = self.unit.get_container(self._container_name)
+        self._oathkeeper_config_path = "/etc/config/oathkeeper.yaml"
+        self._oathkeeper_access_rules_path = "/etc/config/oathkeeper/access-rules.yaml"
+        self._name = self.model.app.name
+
+        self.service_patcher = KubernetesServicePatch(
+            self, [("oathkeeper-api", OATHKEEPER_API_PORT)]
+        )
+
+        self.ingress = IngressPerAppRequirer(
+            self,
+            relation_name="ingress",
+            port=OATHKEEPER_API_PORT,
+            strip_prefix=True,
+        )
+
+        self.framework.observe(self.on.oathkeeper_pebble_ready, self._on_oathkeeper_pebble_ready)
+
+        self.framework.observe(self.ingress.on.ready, self._on_ingress_ready)
+        self.framework.observe(self.ingress.on.revoked, self._on_ingress_revoked)
+
+    @property
+    def _oathkeeper_layer(self) -> Layer:
+        """Returns a pre-configured Pebble layer."""
+        layer_config = {
+            "summary": "oathkeeper-operator layer",
+            "description": "pebble config layer for oathkeeper-operator",
+            "services": {
+                self._container_name: {
+                    "override": "replace",
+                    "summary": "Oathkeeper Operator layer",
+                    "command": f"oathkeeper serve -c {self._oathkeeper_config_path}",
+                    "startup": "enabled",
+                }
+            },
+            "checks": {
+                "alive": {
+                    "override": "replace",
+                    "http": {"url": f"http://localhost:{OATHKEEPER_API_PORT}/health/alive"},
+                },
+                "ready": {
+                    "override": "replace",
+                    "http": {"url": f"http://localhost:{OATHKEEPER_API_PORT}/health/ready"},
+                },
+            },
+        }
+        return Layer(layer_config)
+
+    @property
+    def _oathkeeper_service_is_created(self) -> bool:
+        try:
+            self._container.get_service(self._container_name)
+        except (ModelError, RuntimeError):
+            return False
+        return True
+
+    @property
+    def _kratos_login_url(self) -> str:
```

Comment:
I am still new to the charming, just a tiny question here:

- Why we don't initialize `_kratos_login_url` in the `__init__` method?

Reply:
I'm fine with initializing it in `init` for now, in (near) future there will be a method to take those urls from a kratos-oathkeeper relation interface.